crypto/evp/kdf_lib.c

```diff
@@ -0,0 +1,150 @@
+
+/* This array needs to be in order of NIDs */
+static const EVP_KDF_METHOD *standard_methods[] = {
+    &pbkdf2_kdf_meth,
+#ifndef OPENSSL_NO_SCRYPT
+    &scrypt_kdf_meth,
+#endif
+    &tls1_prf_kdf_meth,
+    &hkdf_kdf_meth
+};
+
+DECLARE_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                           kmeth);
+
+static int kmeth_cmp(const EVP_KDF_METHOD *const *a,
+                     const EVP_KDF_METHOD *const *b)
+{
+    return ((*a)->nid - (*b)->nid);
+}
+
+IMPLEMENT_OBJ_BSEARCH_CMP_FN(const EVP_KDF_METHOD *, const EVP_KDF_METHOD *,
+                             kmeth);
+
+static const EVP_KDF_METHOD *kdf_meth_find(int type)
+{
+    EVP_KDF_METHOD tmp;
+    const EVP_KDF_METHOD *t = &tmp, **ret;
+
+    tmp.nid = type;
+    ret = OBJ_bsearch_kmeth(&t, standard_methods,
+                            OSSL_NELEM(standard_methods));
+    if (ret == NULL || *ret == NULL)
+        return NULL;
+
+    return *ret;
+}
+
+EVP_KDF_CTX *EVP_KDF_CTX_new_id(int id)
+{
+    EVP_KDF_CTX *ret;
+    const EVP_KDF_METHOD *kmeth;
+
+    kmeth = kdf_meth_find(id);
+    if (kmeth == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, EVP_R_UNSUPPORTED_ALGORITHM);
+        return NULL;
+    }
+
+    ret = OPENSSL_zalloc(sizeof(*ret));
+    if (ret == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_NEW_ID, ERR_R_MALLOC_FAILURE);
+        return NULL;
+    }
+
+    if (kmeth->init != NULL && (ret->data = kmeth->init()) == NULL) {
+        EVP_KDF_CTX_free(ret);
+        return NULL;
+    }
+
+    ret->kmeth = kmeth;
+    return ret;
+}
+
+void EVP_KDF_CTX_free(EVP_KDF_CTX *ctx)
+{
+    if (ctx == NULL)
+        return;
+
+    if (ctx->kmeth->cleanup != NULL)
+        ctx->kmeth->cleanup(ctx->data);
+    OPENSSL_free(ctx);
+}
+
+int EVP_KDF_CTX_ctrl(EVP_KDF_CTX *ctx, int cmd, int p1, void *p2)
+{
+    int ret;
+
+    if (ctx == NULL)
+        return 0;
+
+    ret = ctx->kmeth->ctrl(ctx->data, cmd, p1, p2);
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL, EVP_R_COMMAND_NOT_SUPPORTED);
+
+    return ret;
+}
+
+int EVP_KDF_CTX_ctrl_uint64(EVP_KDF_CTX *ctx, int cmd, uint64_t value)
+{
+    return EVP_KDF_CTX_ctrl(ctx, cmd, 0, &value);
+}
+
+int EVP_KDF_CTX_ctrl_str(EVP_KDF_CTX *ctx,
+                         const char *type, const char *value)
+{
+    int ret;
+
+    if (ctx == NULL)
+        return 0;
+
+    if (ctx->kmeth->ctrl_str == NULL) {
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_COMMAND_NOT_SUPPORTED);
+        return -2;
+    }
+
+    ret = ctx->kmeth->ctrl_str(ctx->data, type, value);
+    if (ret == -2)
+        EVPerr(EVP_F_EVP_KDF_CTX_CTRL_STR, EVP_R_COMMAND_NOT_SUPPORTED);
+
+    return ret;
+}
+
+size_t EVP_KDF_CTX_output_size(EVP_KDF_CTX *ctx)
+{
+    if (ctx == NULL || ctx->kmeth->output_size == NULL)
+        return 0;
+
+    return ctx->kmeth->output_size(ctx->data);
+}
+
+int EVP_KDF_CTX_derive(EVP_KDF_CTX *ctx, unsigned char *key, size_t *keylen)
```

Comment:
So @paulidale is talking to me about avoiding the keylen pointer thing and to only have users find out the output size through the dedicated function for it (`EVP_KDF_CTX_output_size` in your case).

Since you and I are current working on fairly similar things and battling the same coding patterns, I think it would be nice if we had some similarity in the resulting API.

I'm up for a discussion for this case.  Pass a pointer to a `size_t` and have it filled, or pass just a `size_t` and have the derivation function return with an error if it's wrong (and is larger than necessary wrong?)

Reply:
Given keylen does not need to convey both buffer size and return (once you have a way to query for output_size if the length can't be arbitrary) then I agree that no passing a pointer is a better API, less chances for mistakes.

Reply:
I got rid of the keylen pointer thing. It is now a size_t.
The way I have implemented it, for fixed-output algorithms (or modes) keylen must exactly match the output size or an error results.